ui/app/scripts/stream/services/metamodel.js

```diff
@@ -0,0 +1,787 @@
+                while ((pos = definitionsText.indexOf('\n', pos)) !== -1) {
+                    linebreaks.push(++pos);
+                }
+                var streamModulesToIds = {}; // foo.bar=99
+
+                // The result should look like this:
+                // {errors:[...], graph:{format:.., streamdefs:[...], nodes:[...], links:[...]}
+
+                var errors = [];
+
+                // streamdefs look like this: {"name":"foo","def":" mail | transform --expression=payload.toString() | log"},
+                var streamdefs = [];
+
+                // nodes look like this: {"id":0,"name":"mail","stream-name":"foo","stream-id":1},
+                var nodes = [];
+
+                // links look like this: {"from":0,"to":1},
+                var links = [];
+
+                var nodeId = 0;
+                var streamNumber = 1;
+                var lineNumber = 0; // which line of text is being processed
+                // streamNumber/lineNumber may not increment at the same rate when there is broken input text
+                while (parsedDefinitions.lines && parsedDefinitions.lines.length !== 0) {
+                    // Process a line
+                    var lineText = definitionsText.substring(linebreaks[lineNumber], linebreaks[lineNumber + 1]);
+                    // Example line: {"errors":null,"success":[{"label":"mail","type":"source","name":"mail","options":{},"sourceChannelName":null,"sinkChannelName":null},{"label":"log","type":"sink","name":"log","options":{},"sourceChannelName":null,"sinkChannelName":null}]}
+                    var line = parsedDefinitions.lines.shift();
+                    var streamName = '';
+                    var streamdef = '';
+                    //var streamStartNodeId = nodeId;
+                    var nameSet = false;
+                    utils.$log.info('convertParseResponseToGraph: Line#' + streamNumber + ': ' + JSON.stringify(line));
+
+                    // Build the graph/links if there was successfully parsed output
+                    var parsedNodes = line.success;
+                    if (parsedNodes) {
+                        var linkFrom = -1;
+                        for (var n = 0; n < parsedNodes.length; n++) {
+                            var linkType = null;
+                            // Example node:
+                            // {"label":"mail","type":"source","name":"mail","options":{},"sourceChannelName":null,"sinkChannelName":null}
+                            var parsedNode = parsedNodes[n];
+                            var graphNode = null;
+                            var channelText;
+                            var newlink;
+                            // check for sourceChannelName
+                            if (parsedNode.sourceChannelName) {
+                                channelText = parsedNode.sourceChannelName;
+                                if (channelText.startsWith('tap:')) { // TAP SOURCE
```

Comment:
I wonder how do we handle the tap source now (with the `channelText` starting with `:` and then the destination name).  May be handled by the next block?